pkg/distributor/influxpush/parser.go

```diff
@@ -0,0 +1,180 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package influxpush
+
+import (
+	"compress/gzip"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"sort"
+	"time"
+
+	io2 "github.com/influxdata/influxdb/v2/kit/io"
+	"github.com/influxdata/influxdb/v2/models"
+	"github.com/prometheus/prometheus/model/labels"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+	"github.com/grafana/mimir/pkg/util"
+)
+
+const internalLabel = "__mimir_source__"
+
+// ParseInfluxLineReader parses a Influx Line Protocol request from an io.Reader.
+func ParseInfluxLineReader(_ context.Context, r *http.Request, maxSize int) ([]mimirpb.TimeSeries, int, error) {
+	qp := r.URL.Query()
+	precision := qp.Get("precision")
+	if precision == "" {
+		precision = "ns"
+	}
+
+	if !models.ValidPrecision(precision) {
+		return nil, 0, fmt.Errorf("precision supplied is not valid: %s", precision)
+	}
+
+	encoding := r.Header.Get("Content-Encoding")
+	reader, err := batchReadCloser(r.Body, encoding, int64(maxSize))
+	if err != nil {
+		return nil, 0, fmt.Errorf("gzip compression error: %w", err)
+	}
+	data, err := io.ReadAll(reader)
+	dataLen := len(data) // In case if something is read despite an error.
+	if err != nil {
+		return nil, dataLen, fmt.Errorf("can't read body: %s", err)
+	}
+
+	err = reader.Close()
+	if err != nil {
+		return nil, dataLen, fmt.Errorf("problem reading body: %s", err)
+	}
+
+	points, err := models.ParsePointsWithPrecision(data, time.Now().UTC(), precision)
+	if err != nil {
+		return nil, dataLen, fmt.Errorf("can't parse points: %s", err)
+	}
+	a, b := writeRequestFromInfluxPoints(points)
+	return a, dataLen, b
+}
+
+func writeRequestFromInfluxPoints(points []models.Point) ([]mimirpb.TimeSeries, error) {
+	// Technically the same series should not be repeated. We should put all the samples for
+	// a series in single client.Timeseries. Having said that doing it is not very optimal and the
+	// occurrence of multiple timestamps for the same series is rare. Only reason I see it happening is
+	// for backfilling and this is not the API for that. Keeping that in mind, we are going to create a new
+	// client.Timeseries for each sample.
+
+	var returnTs []mimirpb.TimeSeries
```

Comment:
This should be allocated to the correct expected size, but also as I commented above, consider returning `[]mimirpb.PreallocTimeseries` here, and taking it from a pool here.